libs/platform/ebpf_epoch.c

```diff
@@ -3,76 +3,62 @@
-// into the _ebpf_epoch_cpu_table[cpu_id].free_list while holding the _ebpf_epoch_cpu_table[cpu_id].free_list_lock.
-// 2) During ebpf_epoch_exit all entries in _ebpf_epoch_cpu_table[cpu_id].free_list where freed_epoch <
-// _ebpf_release_epoch are freed using ebpf_free.
-// 3) During ebpf_epoch_flush the minimum epoch is computed across the values in
-// _ebpf_epoch_cpu_table[*].non_preemptible_epoch and  _ebpf_epoch_cpu_table[*].thread_table[*].entry_epoch (for active
-// threads) and then written to _ebpf_release_epoch.
-// 4) ebpf_epoch_flush is called periodically by _ebpf_flush_timer.
-// 5) ebpf_epoch_flush also queues a non-preemptible work-item (_ebpf_epoch_update_cpu_entry) to all CPUs where
-// _ebpf_epoch_cpu_table[cpu_id].non_preemptible_epoch != _ebpf_current_epoch.
-// 6) _ebpf_epoch_update_cpu_entry calls ebpf_epoch_enter/ebpf_epoch_exit.
+// Third:
+// If the free-list still contains entries, the _ebpf_flush_timer is set (if not already set).
 //
-// Note:
-// CPU table entries aren't cleared on exit as we can't rely on memory ordering.
-// I.e., the thread doing the cleanup may have a stale view of the CPU table.
-// As long as the entries in the CPU table increase, this gives correct behavior.
+// ebpf_flush:
+// Compute the global lowest epoch across all active CPU and thread ebpf_epoch_state_t and set _ebpf_release_epoch.
+//
+// _ebpf_flush_timer:
+// Calls ebpf_flush and clears the _ebpf_flush_timer_set flag.
 //
 
-// Frequency to compute newest inactive epoch.
+// Delay after the _ebpf_flush_timer is set before it runs.
 #define EBPF_EPOCH_FLUSH_DELAY_IN_MICROSECONDS 1000
 
-typedef struct _ebpf_epoch_thread_entry
+typedef struct _ebpf_epoch_state
 {
-    int64_t entry_epoch;
-    int64_t exit_epoch;
-} ebpf_epoch_thread_entry_t;
+    int64_t epoch;
+    bool active;
+} ebpf_epoch_state_t;
 
 // Table to track per CPU state.
+// This table must fit into a multiple of EBPF_CACHE_LINE_SIZE.
 typedef struct _ebpf_epoch_cpu_entry
 {
-    // Discussion: https://github.com/microsoft/ebpf-for-windows/discussions/442
-    // Should this be split into an entry/exit epoch + lock?
-    int64_t non_preemptible_epoch;
-    ebpf_non_preemptible_work_item_t* non_preemptible_work_item;
-    // Discussion: https://github.com/microsoft/ebpf-for-windows/discussions/438
-    // Should this code switch to using an InterlockedSList?
-    ebpf_lock_t free_list_lock;
-    _Requires_lock_held_(free_list_lock) ebpf_list_entry_t free_list;
-    // Discussion: https://github.com/microsoft/ebpf-for-windows/discussions/440
-    // Should this code switch to a lock-free hash table?
-    ebpf_lock_t thread_table_lock;
-    _Requires_lock_held_(thread_table_lock) ebpf_hash_table_t* thread_table;
+    ebpf_lock_t lock;
+    _Requires_lock_held_(lock) ebpf_epoch_state_t cpu_epoch_state;
+    _Requires_lock_held_(lock) ebpf_list_entry_t free_list;
+    _Requires_lock_held_(lock) ebpf_hash_table_t* thread_table;
+    uintptr_t padding[2];
```

Comment:
Leave some comments on how this padding should be adjusted if the struct is changed in future.